File: qbopt/frontend/raising_conditions.py
"""Separate scalar condition reads from their value comparisons."""

from dataclasses import replace

from qbopt.analysis import ssa
from qbopt.model import ir, mir
from qbopt.objectfile.module import Space


def loaded(body: mir.MirBody) -> mir.MirBody:
    values = tuple(ssa.values(body))
    serial = max((value.id for value in values), default=0)
    variable = max((value.variable for value in values), default=0)
    blocks = []
    for block in body.blocks:
        ops = []
        for op in block.ops:
            cells = [arg for arg in op.args if isinstance(arg, mir.Cell)]
            if (op.op is not ir.Operation.COMPARE or op.kind is not mir.Kind.SUB
                or op.results or len(op.defines) != 1 or not op.defines[0].flags
                or op.barrier or op.floating is not None or op.stack is not None or op.merges
                or op.stores or len(op.args) != 2 or len(cells) != 1
                or op.loads != (cells[0].ref,) or cells[0].ref.width not in (2, 4)
                or cells[0].ref.addr is None
                or cells[0].ref.addr.space not in (Space.SEGMENT, Space.FRAME, Space.LITERAL)
                or any(not isinstance(arg, (mir.Cell, mir.Held, mir.Const))
                       or (arg.ref.width if isinstance(arg, mir.Cell) else arg.width)
                       != cells[0].ref.width for arg in op.args)):
                ops.append(op)
                continue
            ref = cells[0].ref
            serial += 1
            variable += 1
            value = mir.Value(serial, op.at, variable=variable, version=1)
            held = mir.Held(value, ref.width)
            ops.append(mir.Op(op.at, ir.Operation.MOVE, "", (value,),
                tuple(part for part in (ref.base, ref.segment) if part is not None),
                kind=mir.Kind.LOAD, loads=(ref,), args=(cells[0],), results=(held,),
                id=next(mir._IDS), symbol=False))
            args = tuple(held if isinstance(arg, mir.Cell) else arg for arg in op.args)
            ops.append(mir.detached(op, args=args, loads=(),
                uses=tuple(dict.fromkeys(arg.value for arg in args if isinstance(arg, mir.Held))),
                raised=None))
        blocks.append(replace(block, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))

File: crates/llrm-core/src/frontends/bc/raising_conditions.rs
//! Port of `qbopt/frontend/raising_conditions.py`.
//!
//! Separate scalar condition reads from their value comparisons.

use crate::analysis::ssa;
use crate::model::ir::Operation;
use crate::model::mir::{self, Arg, Held, Kind, Op, OpCode, RaisedBody, Value};
use crate::objectfile::module::Space;
use crate::support::hash::IndexSet;

pub fn loaded(body: RaisedBody) -> RaisedBody {
    let values: Vec<Value> = ssa::values(&body).collect();
    let mut serial = values.iter().map(|value| value.id).max().unwrap_or(0);
    let mut variable = values.iter().map(|value| value.variable).max().unwrap_or(0);
    let mut blocks = Vec::new();
    for block in &body.blocks {
        let mut ops = Vec::new();
        for op in &block.ops {
            let cells: Vec<_> = op
                .args
                .iter()
                .filter_map(|arg| match arg {
                    Arg::Cell(cell) => Some(cell),
                    _ => None,
                })
                .collect();
            let refused = op.op != Some(OpCode::Operation(Operation::Compare))
                || op.kind != Kind::Sub
                || !op.results.is_empty()
                || op.defines.len() != 1
                || !op.defines[0].flags
                || op.barrier()
                || op.floating.is_some()
                || op.stack.is_some()
                || !op.merges.is_empty()
                || !op.stores.is_empty()
                || op.args.len() != 2
                || cells.len() != 1
                || op.loads != [cells[0].r#ref.clone()]
                || !matches!(cells[0].r#ref.width, 2 | 4)
                || cells[0].r#ref.addr.is_none_or(|addr| !matches!(addr.space, Space::Segment | Space::Frame | Space::Literal))
                || op.args.iter().any(|arg| match arg {
                    Arg::Cell(cell) => cell.r#ref.width != cells[0].r#ref.width,
                    Arg::Held(held) => held.width != cells[0].r#ref.width,
                    Arg::Const(constant) => constant.width != cells[0].r#ref.width,
                    _ => true,
                });
            if refused {
                ops.push(op.clone());
                continue;
            }
            let cell = cells[0].clone();
            let r#ref = cell.r#ref.clone();
            serial += 1;
            variable += 1;
            let value = Value { id: serial, at: op.at, flags: false, variable, version: 1 };
            let held = Held { value, width: r#ref.width };
            let mut load = Op::new(
                op.at,
                OpCode::Operation(Operation::Move),
                "",
                vec![value],
                [r#ref.base, r#ref.segment].into_iter().flatten().collect(),
            );
            load.kind = Kind::Load;
            load.loads = vec![r#ref];
            load.args = vec![Arg::Cell(cell)];
            load.results = vec![Arg::Held(held)];
            load.source = Some(mir::next_id());
            load.symbol = Some(false);
            ops.push(load);
            let args: Vec<Arg> =
                op.args.iter().map(|arg| if matches!(arg, Arg::Cell(_)) { Arg::Held(held) } else { arg.clone() }).collect();
            let mut changed = op.clone();
            changed.uses = args
                .iter()
                .filter_map(|arg| match arg {
                    Arg::Held(held) => Some(held.value),
                    _ => None,
                })
                .collect::<IndexSet<_>>()
                .into_iter()
                .collect();
            changed.args = args;
            changed.loads = Vec::new();
            changed.raised = None;
            ops.push(mir::detached(changed));
        }
        blocks.push(block.with_ops(ops));
    }
    body.with_blocks(blocks)
}
